### PR TITLE
Math package: lcalc-1.23

### PR DESCRIPTION
**Pull request by tornaria**
Michael Rubinstein's L-function calculator.

The long files are just tests. The 3 versions are almost the same except for a few minor rounding differences.

### CODE REVIEW

**Comment by dkwo** [comment]
Did not test it in `sage`, but it builds fine for me.

File: srcpkgs/lcalc/template [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+build_style="gnu-makefile"
+makedepends="pari-devel"
+short_desc="Michael Rubinstein's L-function calculator"
+maintainer="Gonzalo Tornaría <tornaria@cmat.edu.uy>"
+license="GPL-2.0-or-later"
+homepage="https://mirrors.mit.edu/sage/spkg/upstream/lcalc/"
+distfiles="https://mirrors.mit.edu/sage/spkg/upstream/lcalc/lcalc-${version}.tar.bz2"
+checksum=83d4253ec2f38553b21190d6d0c6b71bab7ea14717f6dde5bd18f60775d3cdb0
+patch_args="-Np1"
+nocross=yes
```
Comment (Chocimier) [inline srcpkgs/lcalc/template:15]: Cross builds fine after patching `libLfunction.so: g++` to `libLfunction.so: $(CC)`

File: srcpkgs/lcalc/template [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+pre_install() {
+	mkdir -p ${DESTDIR}/usr/{bin,lib,include}
+	make_install_args+=" INSTALL_DIR=${DESTDIR}/usr"
+}
+
+do_check() {
+	echo "Testing lcalc ..."
+	. "${FILESDIR}/test" > test.log
+	case "$XBPS_TARGET_MACHINE" in
+		i686) 	OUT="${FILESDIR}/test.i686" ;;
```
Comment (Chocimier) [inline srcpkgs/lcalc/template:32]: I don't like hardcoding per architecture, but also don't see easy way to calculate relative error without perl or such.

File: srcpkgs/lcalc/template (hunk 1 above)
Comment (ericonr) [reply srcpkgs/lcalc/template:15]: Shouldn't it be `CXX`, since they are using `g++`?

**Review by ericonr — COMMENTED** [review]
I don't love the idea of the 2k LOC test cases; It'd be better if there was something upstreamable or a program that could generate the testset and compare to what the compiled version outputs.

On the matter of the patches, are they upstream/ going to land upstream? I'm also not in love with the idea of adding the package and us having to maintain patches for any new `pari` change or similar.

File: srcpkgs/lcalc/template (hunk 2 above)
Comment (ericonr) [reply srcpkgs/lcalc/template:32]: Re. my comment below, I'd rather have perl in `hostmakedepends` than the vendored test results (which would also break if the example program changes anything, right?).

**Comment by tornaria** [comment]
General answer to all comments above:
- This program is kind of old and not maintained by the author. It works and it's useful in what it does, but it should be treated mostly as frozen (so, I doubt example programs will change).
- It's included and used in sagemath, which I used as upstream, they will do maintenance if pari changes (I can do it as well).
- patches came mostly from sagemath (https://git.sagemath.org/sage.git/tree/build/pkgs/lcalc/patches)
- I did this in 2018 and I just noticed they have updated patches, in particular there is some makefile cleanup that might be nice to have (like using $(CXX), etc, so cross build will work as noted above).
-  I could update it to have *exactly* the same patches as in sagemath (either included in or as distfiles from a stable tag of the git repo above) and think of sagemath as the upstream.
- The test is just some quick hack I did to know the program worked ok. I can make it smaller if that makes it more palatable. I think at least knowing that the binary runs and that the library can be linked to is good, but maybe instead of computing 1000 zeros of the Riemann zeta function I can just compute 10, etc.

Let me also point out that, other than sagemath, I don't think there will be any dependency of this program. But when we get to package sagemath, this program will have to be compiled, either as a separate package (like here) or as part of the omnibus sagemath.

Also: I rather spend my time and energy on packaging other -- more generally useful -- math software, than polishing too much this (niche) program. All criticism is still very much welcomed, I'm just starting with the "easy" packages.

**Comment by tornaria** [comment]
More on patches: debian uses more or less the same set of patches (https://sources.debian.org/patches/lcalc/1.23+dfsg-11/).

**Comment by ericonr** [comment]
> I could update it to have exactly the same patches as in sagemath (either included in or as distfiles from a stable tag of the git repo above) and think of sagemath as the upstream.

I think the distfiles approach here would be best. Any local patches would go on top of that, and we avoid vendoring in a bunch of "basic" patches.

> But when we get to package sagemath, this program will have to be compiled, either as a separate package (like here) or as part of the omnibus sagemath.

Alternatively, **I** am ok with it being packaged inside of sagemath; it is, after all, the only user for it, so there are very few (if any - maybe it helps with cross builds?) benefits from splitting it into a separate package. Other maintainers might have other opinions on vendoring dependencies, even in this case, so don't consider my opinion the final recommendation.

**Comment by tornaria** [comment]
> I think the distfiles approach here would be best. Any local patches would go on top of that, and we avoid 
vendoring in a bunch of "basic" patches.

Ok. What's a good way to do it? Patch in post_extract() ? Is there something automatic or some template doing that as an example?

> Alternatively, **I** am ok with it being packaged inside of sagemath; it is, after all, the only user for it, so there are very few (if any - maybe it helps with cross builds?) benefits from splitting it into a separate package. Other maintainers might have other opinions on vendoring dependencies, even in this case, so don't consider my opinion the final recommendation.

Well, **I** use it, sometimes. That's why I packaged it for myself a few years ago :-)

**Comment by ericonr** [comment]
The `linux5.11` template does it, for one. We chose to go with `pre_patch` there, but I think `post_extract` is equally valid.

> Well, I use it, sometimes. That's why I packaged it for myself a few years ago :-)

Ah, I guess that's reason enough to make it a system package, then ;)

Would be nice if sage released proper fully patched tarballs, though.

**Comment by tornaria** [comment]
@ericonr 
See what you think.

Summary of changes:
- patches come from sagemath as distfiles, I don't need to patch any further than that.
- test.out is now 112 lines and the same for all architectures
- saner makefile (from new sagemath patches), cross compilation works

**Comment by tornaria** [comment]
> Would be nice if sage released proper fully patched tarballs, though.

I think they used to but they have moved to including patches in the git repo and stocking the upstream tarballs.

File: srcpkgs/lcalc/template [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+post_extract() {
+	for patch in $_patches ; do
+		patch -p1 < "$XBPS_SRCDISTDIR/${pkgname}-${version}/${patch}"
+	done
+}
+
+pre_build() {
+	CXXFLAGS="-fPIC"
+}
```
Comment (ericonr) [inline srcpkgs/lcalc/template:57]: I think you can export this at the template top level just fine.

File: srcpkgs/lcalc/template [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+do_check() {
+	echo "Testing lcalc ..."
+	sh "${FILESDIR}"/test > test.log
+	diff "${FILESDIR}"/test.out test.log && echo PASS
+}
+
+lcalc-devel_package() {
+	short_desc+=" - development files"
+	pkg_install() {
+		vmove "usr/lib"
```
Comment (ericonr) [inline srcpkgs/lcalc/template:73]: Does sage use the binary or the library? If it's the library, I think it should go in the normal package...

File: srcpkgs/lcalc/template [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+ dadf6fc6d481595cb041f8b0dd92bc2e2bc390cbb3e0dd36b7fe63b74942db26
+ 342cde1c9effda48597bfee439acb37c91458a100580c97c84da6a142d2cc5b6
+ 55c6c9f969593f24c75a85e7a150b876671356aa68ceb8e77c563a645897643e
+ 906ef33de15bb0d94883b7797d8cb3c889b9f59bceebeb171e65f1546faf1c37
+ 437e3870342ef4cb5749398ead9381504403e7e1f99efadab95f4a607678b400
+ 6471f2f30497486e289fc8977cbb89c495cf0d5ff58af2b1f479a497a28165e6
+ dde1f80b0ca42fb43917c24e6016c19e4a8d28de9c50cd0a523a318866bdc0eb
+ 941d4ac025f5d6f0142fb785b27f7a221729466dc83d45fd206cce102c012eaf
+ 045f23ea9131b8039001c8f13d22152f7b3c2e91adcf8c555354eaf39e7a7b48
+"
```
Comment (ericonr) [inline srcpkgs/lcalc/template:43]: I'd recommend leaving this together with the initial checksum block; makes it possible for `xgensum` to update everything in one go as well.

File: srcpkgs/lcalc/template [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+ Makefile.patch
+ clang.patch
+ lcalc-1.23_default_parameters_1.patch
+ lcalc-1.23_default_parameters_2.patch
+ pari-2.7.patch
+ pari-mem.patch
+ pari_include.patch
+ time.h.patch
+ using_namespace_std.patch
+"
```
Comment (ericonr) [inline srcpkgs/lcalc/template:30]:
```suggestion
_patches="
 Lcommon.h.patch
 Lvalue.h.patch
 Makefile.patch
 clang.patch
 lcalc-1.23_default_parameters_1.patch
 lcalc-1.23_default_parameters_2.patch
 pari-2.7.patch
 pari-mem.patch
 pari_include.patch
 time.h.patch
 using_namespace_std.patch"
```

Nit

**Comment by ericonr** [comment]
Remember to squash the commits together later ;)

@Chocimier I'd appreciate a second look from you as well.

File: srcpkgs/lcalc/template (hunk 3 above)
Comment (tornaria) [reply srcpkgs/lcalc/template:57]: Indeed.

File: srcpkgs/lcalc/template (hunk 5 above)
Comment (tornaria) [reply srcpkgs/lcalc/template:43]: Sounds good.

File: srcpkgs/lcalc/template (hunk 4 above)
Comment (tornaria) [reply srcpkgs/lcalc/template:72]: I think both. You are right, the `.so` should probably go in the main package. For some reason the binary is compiled static so it doesn't need the `.so` to be around.

**Comment by tornaria** [comment]
WTF!?

The following change makes a rounding difference in i686 -- that's the test failing:
```
-pre_build() {
-    CXXFLAGS="-fPIC -Wno-deprecated"
-}
+CXXFLAGS="-fPIC -Wno-deprecated"
```

**Comment by ericonr** [comment]
I think assigning to `CXXFLAGS` in `pre_build` overrides build flags, including `-O2`. The behavior should be more consistent if you do `CXXFLAGS+=...` in `pre_build`.

**Comment by tornaria** [comment]
> I think assigning to `CXXFLAGS` in `pre_build` overrides build flags, including `-O2`. The behavior should be more consistent if you do `CXXFLAGS+=...` in `pre_build`.

Indeed, `-O2` is making the difference. I found a way to round which makes tests robust.

If you think it's ok to merge, I'll squash.

File: srcpkgs/lcalc/files/test [hunk 7]
```diff
@@ -0,0 +1,17 @@
+echo "### First 1000 zeros of zeta function"
```
Comment (ericonr) [inline srcpkgs/lcalc/files/test:1]: Maybe add a header to this file, explaining that its purpose is to verify both library and executable are working ok, and what the approach to developing the tests was.

**Review by ericonr — COMMENTED** [review]
Overall LGTM, I think it can be squashed

File: srcpkgs/lcalc/files/test (hunk 7 above)
Comment (tornaria) [reply srcpkgs/lcalc/files/test:14]: Done

**Comment by tornaria** [comment]
I wrote an explanation in the test file and squashed all into a single commit.

Thanks.

**Comment by Chocimier** [comment]
One more remark: merging libraries is not promise of merging `sage` if it turn out too ugly. That said, hopefully all will be fine.

**Comment by tornaria** [comment]
@ericonr: rebased with only style changes. Tested with sage-9.4.rc2.

**Comment by dkwo** [comment]
@Chocimier Do you think this can be merged? It seems @ericonr already approved this one, so we can let them rest :)

**Pull request merged**